drivers/rpmsg/rpmsg_gpio.c

```diff
@@ -0,0 +1,190 @@
+ * For example codes to test and run the complete communication between PRU and ARM
+ * https://github.com/deebot/Beaglebone-BidirectionBus/tree/dev
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/gpio.h>
+#include <linux/platform_device.h>
+#include <linux/of.h>
+#include <linux/rpmsg.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/kfifo.h>
+#include <linux/uaccess.h>
+#include <linux/poll.h>
+#include <linux/rpmsg/virtio_rpmsg.h>
+#include <linux/stat.h>
+#include <linux/bitops.h>
+#define PRU_MAX_DEVICES				(16)
+/* Matches the definition in virtio_rpmsg_bus.c */
+#define RPMSG_BUF_SIZE				(512)
+#define MAX_FIFO_MSG				(32)
+#define FIFO_MSG_SIZE				RPMSG_BUF_SIZE
+#define GPIO_NUM 9
+static DEFINE_MUTEX(rpmsg_pru_lock);
+static char rpmsg_pru_buf[RPMSG_BUF_SIZE];
+static struct gpio_chip chip;
+struct rpmsg_pru_dev {
+	struct rpmsg_device *rpdev;
+	struct device *dev;
+	bool locked;
+	dev_t devt;
+	struct kfifo msg_fifo;
+	u32 msg_len[MAX_FIFO_MSG];
+	int msg_idx_rd;
+	int msg_idx_wr;
+	wait_queue_head_t wait_list;
+	uint32_t gpio_state;
+	long input_state ;
+
+};
+
+struct rpmsg_pru_dev *prudev;
+static int mygpio_get_value(struct gpio_chip *gc, unsigned offset){
+	uint32_t value;
+	struct rpmsg_device *rpdev= container_of(gc->parent, struct rpmsg_device, dev);
+	unsigned int mask = BIT(offset % 8);
+	prudev=dev_get_drvdata(&rpdev->dev);
+	value=(prudev->input_state& mask)>>offset;
+	return value;
+}
+
+static void mygpio_set_value(struct gpio_chip *gc, unsigned offset, int val)
+{
+	int ret;
+	struct rpmsg_device *rpdev= container_of(gc->parent, struct rpmsg_device, dev);
+	pr_info("set_value function triggered");
+	pr_info("The bit number %d is set to value: %d",offset,val);
+
+	prudev = dev_get_drvdata(&rpdev->dev);
+	if (val==0){
+		prudev->gpio_state &= ~(1<<offset);
+		//pr_info("A when value 0: %d",~(1<<offset));
+	}
+	else{
+		prudev->gpio_state |= (1<<offset);
+		//pr_info("A when value 1: %d",(1<<offset));
+	}
+	if(offset==GPIO_NUM-1){
+		memcpy((void*)&rpmsg_pru_buf,(void*)&(prudev->gpio_state),sizeof(&(prudev->gpio_state)));
+		pr_info("A check for checking gpio_state: %d",prudev->gpio_state);
+		ret=rpmsg_send(rpdev->ept, (void *)rpmsg_pru_buf, 2);
+		if (ret)
+			dev_err(gc->parent, "rpmsg_send failed: %d\n", ret);
+	}
+
+}
+
+static int mygpio_direction_output(struct gpio_chip *gc,
+				       unsigned offset, int val)
+{
+	printk("Direction of GPIO set to: out\n");
```

Comment:
Provide some comments on what work remains to be done. This looks like just a stub. Is it needed in the future?